### PR TITLE
feat: Auto Form component

### PR DESCRIPTION
**Pull request by Artur-**
Can be used as
```
<ExperimentalAutoForm service={AppointmentService} model={AppointmentModel} />
```

to auto generate a form using AppointmentModel and allow submitting the result to the update method of the AppointmentService

### CODE REVIEW

File: packages/ts/react-grid/src/autoform.tsx [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+              setFormError('');
+              const newItem = await form.submit();
+              if (!newItem) {
+                // If update returns an empty object, then no update was performed
+                throw new Error('generic error');
+              } else if (onSubmit) {
+                onSubmit({ item: newItem });
+              } else {
+                form.clear();
+              }
```
Comment (sissbruecker) [inline packages/ts/react-grid/src/autoform.tsx:60]: Don't quite understand this part: If there is a callback, then we don't clear the form, otherwise we do.

I think the form should be cleared when:
- There is no `item` and the form has been successfully submitted
- The `item` state changes to null. Might already be covered by `form.read`, not sure though.

File: packages/ts/react-grid/src/autoform.tsx [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+  const [formError, setFormError] = useState('');
+  useEffect(() => {
+    form.read(item);
+  }, [item]);
+
+  return (
+    <VerticalLayout theme="padding spacing">
+      {getProperties(model).map((propertyInfo) => (
+        <AutoFormField
+          key={`auto-form-field-${propertyInfo.name}`}
```
Comment (sissbruecker) [inline packages/ts/react-grid/src/autoform.tsx:38]:
```suggestion
          key={propertyInfo.name}
```
Nit: The key is never rendered into the DOM, it just needs to be a unique value. As such we can remove any constants from it as those don't make it any more unique.

File: packages/ts/react-grid/test/autoform.spec.tsx [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+        someNumber: 24451,
+        vip: false,
+      };
+      const result = render(<AutoForm service={personService} model={PersonModel} item={person} />);
+      await assertFormFieldValue(result, 'First name', person.firstName);
+      await assertFormFieldValue(result, 'Last name', person.lastName);
+      await assertFormFieldValue(result, 'Email', person.email);
+      await assertFormFieldValue(result, 'Some number', person.someNumber);
+    });
+    it('works without an item', async () => {
```
Comment (sissbruecker) [inline packages/ts/react-grid/test/autoform.spec.tsx:67]: Could add a test that the form gets updated when passing a different item or switching from an item to null.

File: packages/ts/react-grid/test/autoform.spec.tsx [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+      const service: CrudService<Person> & HasLastFilter = createService<Person>(personData);
+
+      const person = await getItem(service, 1);
+      let submittedItem: Person | undefined;
+      const result = render(
+        <AutoForm
+          service={service}
+          model={PersonModel}
+          item={person}
+          onSubmit={({ item }) => {
```
Comment (sissbruecker) [inline packages/ts/react-grid/test/autoform.spec.tsx:106]: Nit: Could use a spy here.

File: packages/ts/react-grid/test/autoform.spec.tsx
Comment (sissbruecker) [inline packages/ts/react-grid/test/autoform.spec.tsx:1]: Clearing the form is lacking tests.

File: packages/java/endpoint/src/main/java/dev/hilla/crud/CrudService.java [hunk 5]
```diff
@@ -26,4 +26,25 @@ public interface CrudService<T> {
...
+     * <p>
+     * If you store the object in a SQL database, the returned object might have
+     * a new id or updated consistency version.
+     *
+     * @param value
+     *            the object to update/store
+     * @return the fresh object or {@code null} if no object was found to update
+     */
+    @Nullable
+    T update(T value);
```
Comment (sissbruecker) [inline packages/java/endpoint/src/main/java/dev/hilla/crud/CrudService.java:41]: How about using `save` here? Update makes it look like there should be a `create`, and you need to check the API docs to verify that this can also create items. `Save` is already familiar terminology from JPA.

File: packages/ts/react-grid/test/autoform.spec.tsx [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+    it('uses values from an item', async () => {
+      const person = (await getItem(personService, 2))!;
+
+      const result = render(<AutoForm service={personService} model={PersonModel} item={person} />);
+      await assertFormFieldValue(result, 'First name', person.firstName);
+      await assertFormFieldValue(result, 'Last name', person.lastName);
+      await assertFormFieldValue(result, 'Email', person.email);
+      await assertFormFieldValue(result, 'Some number', person.someNumber);
+    });
+    it('submits a valid form', async () => {
```
Comment (sissbruecker) [inline packages/ts/react-grid/test/autoform.spec.tsx:83]: Could also test creating a new item.

File: packages/ts/react-grid/test/autoform.spec.tsx [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+      };
+      const person = await getItem(service, 1);
+      const result = render(<AutoForm service={service} model={PersonModel} item={person} />);
+      await submit(result);
+      await nextFrame();
+      // eslint-disable-next-line
+      const formError = (result.container.querySelector('#formerror') as HTMLElement).innerText;
+      expect(formError).to.equal('Something went wrong, please check all your values');
+    });
+    it('calls onSubmitEerror and does not show error if the endpoint call fails', async () => {
```
Comment (sissbruecker) [inline packages/ts/react-grid/test/autoform.spec.tsx:130]: Could also test that `onSubmit` is not called when there is an error.

File: packages/java/tests/spring/react-grid-test/src/test/java/dev/hilla/test/reactgrid/AutoFormViewIT.java [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+        getTextField("doctor").setValue("Evil");
+        getTextField("time").setValue("now");
+        submit();
+
+        // Field values remain
+        Assert.assertEquals("It's me, Mario", getTextField("name").getValue());
+        // Custom error message shown
+        Assert.assertEquals(
+                "Something went wrong, please check all your values",
+                $("div").id("formerror").getText());
```
[the remark this answers is not in the document]
Comment (Artur-) [reply packages/java/tests/spring/react-grid-test/src/test/java/dev/hilla/test/reactgrid/AutoFormViewIT.java]: Removed

File: packages/ts/react-grid/test/autoform.spec.tsx (hunk 3 above)
Comment (Artur-) [reply packages/ts/react-grid/test/autoform.spec.tsx:71]: Added

File: packages/ts/react-grid/test/autoform.spec.tsx (hunk 7 above)
Comment (Artur-) [reply packages/ts/react-grid/test/autoform.spec.tsx]: Fixed

File: packages/ts/react-grid/test/autoform.spec.tsx (hunk 4 above)
Comment (Artur-) [reply packages/ts/react-grid/test/autoform.spec.tsx:165]: Done

File: packages/ts/react-grid/test/autoform.spec.tsx
Comment (Artur-) [reply packages/ts/react-grid/test/autoform.spec.tsx:1]: Added

File: packages/ts/react-grid/src/autoform.tsx (hunk 1 above)
Comment (Artur-) [reply packages/ts/react-grid/src/autoform.tsx]: Yeah the thought was that if you handle submit yourself you might also want to control what happens to the form. But maybe it always should be cleared.

File: packages/ts/react-grid/src/autoform.tsx [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+          form={form}
+        ></AutoFormField>
+      ))}
+      <div id="formerror" style={{ color: 'red' }}>
+        {formError}
+      </div>
+      <HorizontalLayout theme="spacing">
+        <Button
+          // eslint-disable-next-line @typescript-eslint/no-misused-promises
+          onClick={async () => {
```
[the remark this answers is not in the document]
Comment (Artur-) [reply packages/ts/react-grid/src/autoform.tsx]: Done

File: packages/java/endpoint/src/main/java/dev/hilla/crud/CrudService.java (hunk 5 above)
Comment (Artur-) [reply packages/java/endpoint/src/main/java/dev/hilla/crud/CrudService.java:41]: That would maybe fit better, I will change it

Comment (Artur-) [reply packages/java/endpoint/src/main/java/dev/hilla/crud/CrudService.java]: Done

File: packages/ts/react-grid/src/autoform.tsx (hunk 1 above)
Comment (sissbruecker) [reply packages/ts/react-grid/src/autoform.tsx]: Always clearing might be a problem when implementing the CRUD view, I would expect that a selected item stays in the form after saving. But we can also deal with this later.

File: packages/ts/react-grid/test/autoform.spec.tsx (hunk 4 above)
Comment (sissbruecker) [reply packages/ts/react-grid/test/autoform.spec.tsx:148]: I think you missed to update it in this specific test.

File: packages/ts/react-grid/test/autoform.spec.tsx (hunk 6 above)
Comment (sissbruecker) [reply packages/ts/react-grid/test/autoform.spec.tsx:104]: It's probably OK to leave this out.

File: packages/ts/react-grid/test/autoform.spec.tsx (hunk 4 above)
Comment (Artur-) [reply packages/ts/react-grid/test/autoform.spec.tsx:148]: done

**Review by sissbruecker — APPROVED** [review]
